cmd/scollector/collectors/aws.go

```diff
@@ -0,0 +1,314 @@
+	var md opentsdb.MultiDataPoint
+	creds := aws.Creds(accessKey, secretKey, "")
+	if creds == nil {
+		return nil, fmt.Errorf("unable to make creds")
+	}
+	ecc := ec2.New(creds, region, nil)
+	if ecc == nil {
+		return nil, fmt.Errorf("unable to login to EC2")
+	}
+	elb := elb.New(creds, region, nil)
+	if elb == nil {
+		return nil, fmt.Errorf("unable to login to ELB")
+	}
+	cw := cloudwatch.New(creds, region, nil)
+	if cw == nil {
+		return nil, fmt.Errorf("unable to login to CloudWatch")
+	}
+	instances, err := AWSGetInstances(*ecc)
+	if err != nil {
+		slog.Info("No EC2 Instances found.")
+	}
+	loadBalancers, err := AWSGetLoadBalancers(*elb)
+	if err != nil {
+		slog.Info("No ELB Load Balancecrs found.")
+	}
+	for _, loadBalancer := range loadBalancers {
+		AWSGetELBLatency(*cw, &md, loadBalancer)
+		AWSGetELBHostCounts(*cw, &md, loadBalancer)
+	}
+	for _, instance := range instances {
+		AWSGetCPU(*cw, &md, instance)
+		AWSGetNetwork(*cw, &md, instance)
+		AWSGetDiskBytes(*cw, &md, instance)
+		AWSGetDiskOps(*cw, &md, instance)
+		AWSGetStatusChecks(*cw, &md, instance)
+	}
+	return md, nil
+}
+
+func AWSGetInstances(ecc ec2.EC2) ([]ec2.Instance, error) {
+	instancelist := []ec2.Instance{}
+	resp, err := ecc.DescribeInstances(nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to describe EC2 Instances")
+	}
+	for _, reservation := range resp.Reservations {
+		for _, instance := range reservation.Instances {
+			instancelist = append(instancelist, instance)
+		}
+	}
+	return instancelist, nil
+}
+
+func AWSGetLoadBalancers(lb elb.ELB) ([]elb.LoadBalancerDescription, error) {
+	lbList := []elb.LoadBalancerDescription{}
+	resp, err := lb.DescribeLoadBalancers(nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to describe ELB Balancers")
+	}
+	for _, loadBalancer := range resp.LoadBalancerDescriptions {
+		lbList = append(lbList, loadBalancer)
+	}
+	return lbList, nil
+}
+
+func AWSGetCPU(cw cloudwatch.CloudWatch, md *opentsdb.MultiDataPoint, instance ec2.Instance) error {
+	search := cloudwatch.GetMetricStatisticsInput{
+		StartTime:  time.Now().UTC().Add(time.Second * -600),
+		EndTime:    time.Now().UTC(),
+		MetricName: aws.String("CPUUtilization"),
+		Period:     aws.Integer(60),
+		Statistics: []string{"Average"},
+		Namespace:  aws.String("AWS/EC2"),
+		Unit:       aws.String("Percent"),
+		Dimensions: []cloudwatch.Dimension{{Name: aws.String("InstanceId"), Value: instance.InstanceID}},
+	}
+	resp, err := cw.GetMetricStatistics(&search)
+	if err != nil {
+		return fmt.Errorf("error getting Metric Statistics: %s", err)
+	}
+	tags := opentsdb.TagSet{
+		"instance": *instance.InstanceID,
+	}
+	for _, datapoint := range resp.Datapoints {
+		AddTS(md, awsCPU, datapoint.Timestamp.Unix(), *datapoint.Average, tags, metadata.Gauge, metadata.Pct, descAWSEC2CPU)
+	}
+	return nil
+}
+func AWSGetNetwork(cw cloudwatch.CloudWatch, md *opentsdb.MultiDataPoint, instance ec2.Instance) error {
```

Comment:
These and some of the following functions seem to share a lot of code - I bet you can refactor these into a generic function where you just pass the Unit and maybe the metadata / tag information. Or at least a func to initialize cloudwatch.GetMetricStatisticsInput since most of the values are the same.